File: 31.下一个排列.py
#
# @lc app=leetcode.cn id=31 lang=python3
#
# [31] 下一个排列
#
# https://leetcode-cn.com/problems/next-permutation/description/
#
# algorithms
# Medium (36.85%)
# Likes:    1154
# Dislikes: 0
# Total Accepted:    171.3K
# Total Submissions: 463.3K
# Testcase Example:  '[1,2,3]'
#
# 实现获取 下一个排列 的函数，算法需要将给定数字序列重新排列成字典序中下一个更大的排列。
# 
# 如果不存在下一个更大的排列，则将数字重新排列成最小的排列（即升序排列）。
# 
# 必须 原地 修改，只允许使用额外常数空间。
# 
# 
# 
# 示例 1：
# 
# 
# 输入：nums = [1,2,3]
# 输出：[1,3,2]
# 
# 
# 示例 2：
# 
# 
# 输入：nums = [3,2,1]
# 输出：[1,2,3]
# 
# 
# 示例 3：
# 
# 
# 输入：nums = [1,1,5]
# 输出：[1,5,1]
# 
# 
# 示例 4：
# 
# 
# 输入：nums = [1]
# 输出：[1]
# 
# 
# 
# 
# 提示：
# 
# 
# 1 
# 0 
# 
# 
#

# @lc code=start
class Solution:
    def findFirstGreater(self, nums: list[int], target: int, left: int, right: int):
        while left + 1 < right:
            middle = math.floor((left + right) / 2)
            if nums[middle] <= target:
                right = middle
            else:
                left = middle
        if nums[right] > target: return right
        return left

    def nextPermutation(self, nums: list[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in range(len(nums) - 2, -1, -1):
            if nums[i] < nums[i + 1]: break
        firstGreater = self.findFirstGreater(nums, nums[i], i + 1, len(nums) - 1)
        nums[i], nums[firstGreater] = nums[firstGreater], nums[i]
        nums[i + 1:] = sorted(nums[i + 1:])
# @lc code=end

# nums = [1,3,2]
# Solution().nextPermutation(nums)
# print(nums)

File: 31.下一个排列.ts
/*
 * @lc app=leetcode.cn id=31 lang=typescript
 *
 * [31] 下一个排列
 *
 * https://leetcode.cn/problems/next-permutation/description/
 *
 * algorithms
 * Medium (37.90%)
 * Likes:    1869
 * Dislikes: 0
 * Total Accepted:    351.2K
 * Total Submissions: 926.5K
 * Testcase Example:  '[1,2,3]'
 *
 * 整数数组的一个 排列  就是将其所有成员以序列或线性顺序排列。
 *
 *
 * 例如，arr = [1,2,3] ，以下这些都可以视作 arr 的排列：[1,2,3]、[1,3,2]、[3,1,2]、[2,3,1] 。
 *
 *
 * 整数数组的 下一个排列 是指其整数的下一个字典序更大的排列。更正式地，如果数组的所有排列根据其字典顺序从小到大排列在一个容器中，那么数组的 下一个排列
 * 就是在这个有序容器中排在它后面的那个排列。如果不存在下一个更大的排列，那么这个数组必须重排为字典序最小的排列（即，其元素按升序排列）。
 *
 *
 * 例如，arr = [1,2,3] 的下一个排列是 [1,3,2] 。
 * 类似地，arr = [2,3,1] 的下一个排列是 [3,1,2] 。
 * 而 arr = [3,2,1] 的下一个排列是 [1,2,3] ，因为 [3,2,1] 不存在一个字典序更大的排列。
 *
 *
 * 给你一个整数数组 nums ，找出 nums 的下一个排列。
 *
 * 必须 原地 修改，只允许使用额外常数空间。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [1,2,3]
 * 输出：[1,3,2]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = [3,2,1]
 * 输出：[1,2,3]
 *
 *
 * 示例 3：
 *
 *
 * 输入：nums = [1,1,5]
 * 输出：[1,5,1]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1 <= nums.length <= 100
 * 0 <= nums[i] <= 100
 *
 *
 */

export
// @lc code=start
function nextPermutation(nums: number[]): void {
  let p = nums.length - 1
  while (p - 1 >= 0 && nums[p - 1] >= nums[p]) {
    p--
  }
  if (p === 0) {
    nums.reverse()
    return
  }

  p--
  let i = nums.length - 1
  while (nums[i] <= nums[p]) {
    i--
  }

  swap(nums, i, p)
  reverse(nums, p + 1, nums.length - 1)
}

function swap(arr: any[], i: number, j: number) {
  const temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
}

function reverse(arr: any, i: number, j: number) {
  while (i < j) {
    swap(arr, i++, j--)
  }
}
// @lc code=end
